<div class="alert alert-success">
    
**Комментарий ревьюера**

Привет! Меня зовут Сысоева Вера.
    
Далее в файле мои комментарии ты сможешь найти в ячейках, аналогичных данной ( если рамки комментария зелёные и отмечены **#ок** - всё сделано правильно; жёлтые и отмечены **#можно лучше** - есть замечания, но не критично; красные с пометкой **#ошибка** - нужно переделать). Не удаляй эти комментарии и постарайся учесть их в ходе выполнения проекта.
    
    
Я буду поддерживать версионность комментариев, и при следующих итерациях комментарии будут выделены, как "Комментраии ревьюера - 2 (3 и.т.д.)"
    
Если у тебя есть замечания, возражения и вопросы, ты можешь писать свои комментарии, по возможности выделяй их так же цветом (например, синим), чтоб их можно было легко отделить от текста проекта. 
    
Кстати, предлагаю сразу общаться на "ты".
    
Увидев у тебя ошибку, в первый раз я лишь укажу на ее наличие и дам тебе возможность самостоятельно найти и исправить ее. На реальной работе твой начальник будет поступать так же, а я пытаюсь подготовить тебя именно к работе аналитиком. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку.
    
</div>

Компанией было приобретен крупный сервис для чтения книг по подписке. Необходимо проанализировать имеющуюся информацию и определить направления развития

- Цель:

Пронализировать базу данных, в которой находятся данные о книгах с целью формулирования ценностного предложения для нового продукта.

- Исходные данные:

Информация о книгах, издательствах, авторах, а также пользовательские обзоры книг

База данных, содержащая таблицы: *books, authors, publishers, ratings и reviews

In [1]:
# импортируем библиотеки

import pandas as pd
from sqlalchemy import create_engine
import warnings
warnings.filterwarnings("ignore")

In [29]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [ ]:
!pip install psycopg2

In [ ]:
query = '''
           SELECT *
           FROM books
        '''

books = pd.io.sql.read_sql(query, con=engine)

In [ ]:
books.head(5)

    Таблица books содержит 1000 записей. В таблице представлены идентификаторы книг, авторов и издательств, а также 
    наименования книг, количество страниц и дату издания.

In [57]:
query = '''
           SELECT *
           FROM authors
        '''

authors = pd.io.sql.read_sql(query, con=engine)

In [58]:
authors.head(5)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [ ]:
query = '''
           SELECT COUNT(author_id)
           FROM authors
        '''

authors2 = pd.io.sql.read_sql(query, con=engine)

In [ ]:
authors2.head(5)

     Таблица authors содержит 636 записей об авторах книг. Таблица содержит идентификаторы авторов и имена/псевдонимы. По 
     первым строкам таблицы можно увидеть упорядоченость по алфавиту

In [10]:
query = '''
           SELECT COUNT(publisher_id)
           FROM publishers
        '''

publishers = pd.io.sql.read_sql(query, con=engine)

In [11]:
publishers.head(5)

count
0    340

In [12]:
query = '''
           SELECT *
           FROM publishers
        '''

publishers2 = pd.io.sql.read_sql(query, con=engine)

In [13]:
publishers2.head(5)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

    Таблица publishers содержит 340 записей об издательствах: идентификаторы издательств и названия. Аналогично по первым 
    строкам таблицы можно увидеть упоРядоченость по алфавиту

In [14]:
query = '''
           SELECT *
           FROM ratings
        '''

ratings = pd.io.sql.read_sql(query, con=engine)

In [15]:
ratings.head(5)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [16]:
query = '''
           SELECT COUNT(rating_id)
           FROM ratings
        '''

ratings2 = pd.io.sql.read_sql(query, con=engine)

In [17]:
ratings2.head(5)

count
0   6456

    Таблица ratings содержит 6456 записей. Таблица содержит информацию об идентификаторах книг и оценки, а также имя 
    пользователя и поставленная этим пользователем оценка.

In [18]:
query = '''
           SELECT *
           FROM reviews
        '''

reviews = pd.io.sql.read_sql(query, con=engine)

In [19]:
reviews.head(5)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [20]:
query = '''
           SELECT COUNT(review_id)
           FROM reviews
        '''

reviews2 = pd.io.sql.read_sql(query, con=engine)

In [21]:
reviews2.head(5)

count
0   2793

    Таблица reviews содержит 2793 записей. Таблица содержит информацию об идентификаторах книг и обзоров, а также имя 
    пользователя и текст обзора.

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [22]:
query = '''
           SELECT count(book_id)
           FROM books
           WHERE publication_date > '2000-01-01'
        '''

q1 = pd.io.sql.read_sql(query, con=engine)

In [23]:
print(q1)

   count
0    819


    Начиная со 2 января 2000 года было издано 819 книг.

<div class="alert alert-success">
    
**Комментарий ревьюера**

**#ок**

Верно
</div>

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [24]:
query = '''
           SELECT DISTINCT b.book_id,
                  b.title,
                  ROUND(AVG(r.rating) OVER (PARTITION BY b.book_id), 2) AS avg_rating,
                  COUNT(r.rating_id) OVER (PARTITION BY b.book_id) AS qnt_of_reviews
           FROM books as b
           JOIN ratings as r ON b.book_id = r.book_id
           ORDER BY avg_rating DESC
        '''

q2 = pd.io.sql.read_sql(query, con=engine)

In [25]:
display(q2)

book_id                                              title  avg_rating  \
0         17                      A Dirty Job (Grim Reaper  #1)        5.00   
1         20              A Fistful of Charms (The Hollows  #4)        5.00   
2         55          A Woman of Substance (Emma Harte Saga #1)        5.00   
3         57                    Act of Treason (Mitch Rapp  #9)        5.00   
4         62                                      Alas  Babylon        5.00   
..       ...                                                ...         ...   
995      915  The World Is Flat: A Brief History of the Twen...        2.25   
996      202                                      Drowning Ruth        2.00   
997      316                  His Excellency: George Washington        2.00   
998      371                                              Junky        2.00   
999      303                               Harvesting the Heart        1.50   

     qnt_of_reviews  
0                 4  
1                 2  
2                 2  
3                 2  
4                 2  
..              ...  
995               4  
996               3  
997               2  
998               2  
999               2  

[1000 rows x 4 columns]

    Для каждой книги мы определили количество обзоров и среднюю оценку. Отсортировав данные можно заметить, что максимально
    поставленной оценкой является 5, а минимально поставленной 1.5.

<div class="alert alert-success">
    
**Комментарий ревьюера**

**#ок**

Да, правильно, что учтена каждая книга
</div>

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [26]:
query = '''
           SELECT DISTINCT p.publisher,
                  COUNT(b.book_id) OVER (PARTITION BY b.publisher_id) AS sum_books
           FROM publishers as p
           JOIN books as b ON p.publisher_id = b.publisher_id
           WHERE b.num_pages > 50 
           ORDER BY sum_books DESC
           LIMIT 1
        '''

q3 = pd.io.sql.read_sql(query, con=engine)

In [27]:
print(q3)

       publisher  sum_books
0  Penguin Books         42


    Больше всего книг (42 штуки) объемом более 50 страниц выпустило издательство Penguin Books.

<div class="alert alert-success">
    
**Комментарий ревьюера**

**#ок**

Верный результат
</div>

### Определите автора с самой высокой средней оценкой книг — только книги с 50 и более оценками

In [59]:
query = '''
           SELECT a.author,
       AVG(r.rating) avg_rating
FROM authors AS a
JOIN books AS b ON a.author_id=b.author_id
JOIN ratings AS r ON b.book_id=r.book_id
WHERE b.book_id IN
    (SELECT b.book_id
     FROM books AS b
     JOIN ratings AS r ON b.book_id=r.book_id
     GROUP BY b.book_id
     HAVING count(r.rating_id)>=50)
GROUP BY a.author
ORDER BY avg_rating desc

LIMIT 1
           
        '''

q4 = pd.io.sql.read_sql(query, con=engine)

In [60]:
print(q4) 

                       author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097


    Книгой с самой высокой средней оценкой равной 4.287 является книга Joanne Rowling.

<div class="alert alert-danger">
    
**Комментарий ревьюера**

**#ошибка**

Автор верный, а средний рейтинг нет. Здесь нужно сделать чуть более ступенчатый запрос. Сначала определи книги с 50 обзорами и более. Потом только для этих книг присоедини автора, и потом посчитай средний рейтинг
</div>

<div class="alert alert-info"> <b>Комментарий студента:</b> Исправила. </div>

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [ ]:
query = '''
           WITH users_more_50 AS
  (SELECT r.username
   FROM ratings as r
   GROUP BY r.username
   HAVING (count(r.rating_id) > 50))
SELECT round(avg(t.cnt)) AS avg_review
FROM
  (SELECT count(rev.review_id) AS cnt
   FROM reviews rev
   WHERE rev.username in
       (SELECT *
        FROM users_more_50)
   GROUP BY rev.username) AS t
           
        '''

q5 = pd.io.sql.read_sql(query, con=engine)

In [ ]:
print(q5)

    В среднем, пользователь, поставивший больше 50-ти оценок составляет 24 обзора на книги.

<div class="alert alert-success">
    
**Комментарий ревьюера**

**#ок**
    
+

</div>

### Вывод

    Для каждой книги было посчитано количество обзоров и средняя оценка. Можно предположить, что чем больше обзоров - 
    - тем достовернее оценка. 
    Возможно, стоит выделить в приложении раздел с топовыми по рейтингу и количеству обзоров книгами, чтобы пользователь
    видел самые популярные книги при открытии сервиса.

<div class="alert alert-success">
    
**Комментарий ревьюера**

**#ок**
    
Теперь мы знаем многие операции с `SQL` в использовании `python` и это отлично, потому что мы будем достаточно часто использовать в нашей практике.
    
Полезные ссылки по SQL:
    
https://habr.com/ru/post/268983/
    
https://habr.com/ru/post/340460/

</div>